tfhe/src/core_crypto/gpu/vec.rs

```diff
@@ -23,30 +24,72 @@ use tfhe_cuda_backend::cuda_bind::{
 /// memory, it is pretty close to a `Vec`. That being said, it only present a very very limited api.
 #[derive(Debug)]
 pub struct CudaVec<T: Numeric> {
-    ptr: CudaPtrMut,
-    len: usize,
-    gpu_index: u32,
+    pub ptr: Vec<*mut c_void>,
+    pub lengths: Vec<usize>,
+    pub gpu_indexes: Vec<u32>,
     _phantom: PhantomData<T>,
 }
 
 impl<T: Numeric> CudaVec<T> {
-    pub fn new(len: usize, streams: &CudaStreams) -> Self {
-        let vec = unsafe { Self::new_async(len, streams) };
+    pub fn new(len: usize, streams: &CudaStreams, gpu_index: u32) -> Self {
+        let vec = unsafe { Self::new_async(len, streams, gpu_index) };
         streams.synchronize();
         vec
     }
     /// # Safety
     ///
     /// - `streams` __must__ be synchronized to guarantee computation has finished
-    pub unsafe fn new_async(len: usize, streams: &CudaStreams) -> Self {
+    pub unsafe fn new_async(len: usize, streams: &CudaStreams, gpu_index: u32) -> Self {
         let size = len as u64 * std::mem::size_of::<T>() as u64;
-        let ptr = cuda_malloc_async(size, streams.ptr[0], streams.gpu_indexes[0]);
-        cuda_memset_async(ptr, 0u64, size, streams.ptr[0], streams.gpu_indexes[0]);
+        let ptr = cuda_malloc_async(
+            size,
+            streams.ptr[gpu_index as usize],
+            streams.gpu_indexes[gpu_index as usize],
+        );
+        cuda_memset_async(
+            ptr,
+            0u64,
+            size,
+            streams.ptr[gpu_index as usize],
+            streams.gpu_indexes[gpu_index as usize],
+        );
+
+        Self {
+            ptr: vec![ptr; 1],
+            lengths: vec![len; 1],
+            gpu_indexes: vec![streams.gpu_indexes[gpu_index as usize]; 1],
+            _phantom: PhantomData,
+        }
+    }
+
+    pub fn new_multi_gpu(len: usize, streams: &CudaStreams) -> Self {
```

Comment:
Add a docstring saying that this creates a CudaVec that holds memory of `len` elements on _each_ GPU that the `CudaStreams` are on